### ISSUE TITLE
fit with SmoothBrokenPowerLaw issue

### ISSUE THREAD

**Issue by dmalishev** [fermi-lat/Likelihood#54]
I've noticed, that a fit with SmoothBrokenPowerLaw model leads to clearly wrong best-fit parameters even for relatively bright sources. The best-fit model does not pass through the spectral points (although the initial parameters seems to describe the data relatively well) and best-fit values for e.g. slope Index1 seems always hit the minimal allowed value -5.

I've checked that the flux calculated according to this model looks fine (the model flux describes the data well for reasonable parameters values). I presume thus that the issue could be due to the wrong derivatives (over SmoothBrokenPowerLaw parameters) calculation.

May you check this issue, please?

**Comment by nmirabal** [comment]
Some advice from Jean and Marianne. The SBPL convergence is tricky, especially if you have a large number of degrees of freedom. It is very rare that beta can actually be fitted. The first solution to try is to fix beta (to 0.2, for example). Setting Scale is also delicate, it should be close to the pivot energy at which the error on flux density is not too large. Having Index1 hit -5 indicates that the model has inverted curvature and indicates an issue. There are a number of factors here and in general the log parabola often converge more easily than the SBPL.

**Comment by dmalishev** [comment]
Unfortunately fixing beta did not help. I was trying to mimic a simple powerlaw with the broken one by fixing beta, index2 and BreakValue, but still got convergence to non-correct values, please see "best-fit" below

` <spectrum type="SmoothBrokenPowerLaw">
      <parameter error="4814645.412" free="1" max="1e+10" min="0" name="Prefactor" scale="1e-06" value="244726617.7" />
      <parameter error="0.03192203655" free="1" max="-1" min="-5" name="Index1" scale="1" value="-4.937832271" />
      <parameter free="0" max="2000" min="30" name="Scale" scale="1" value="100" />
      <parameter free="0" max="-1" min="-5" name="Index2" scale="1" value="-2" />
      <parameter free="0" max="500000" min="20" name="BreakValue" scale="1" value="400000" />
      <parameter free="0" max="10" min="0.01" name="Beta" scale="1" value="0.2" />
    </spectrum>`

**Comment by nmirabal** [comment]
Having Index1 approach -5 means that the model has inverted curvature, so there is an issue here. It is typically more common to have the model hit the upper limit of -1 rather than the lower limit. Have you tried a log parabola fit?.

**Comment by dmalishev** [comment]
Yes, I've tried it previously. Everything works fine with LogParabola and a PowerLaw fit

**Comment by dmalishev** [comment]
Please find results of PL and LogParabola fit of the same data attached. Both look ok to me.
[pl_lp_fit.txt](https://github.com/fermi-lat/Likelihood/files/4255284/pl_lp_fit.txt)

**Comment by nmirabal** [comment]
It is difficult to add more here as the SmoothBrokenPowerLaw function is so finicky. Two final suggestions a) Try to set Scale as close to the break as possible and leave it fixed, b) start experimenting with changing beta between 0.01 and 10 in small increments and again leaving it fixed. With many so many free parameters, it is very rare that SmoothBrokenPowerLaw can actually be fitted.

**Comment by Areustle** [comment]
We'll need to ensure the gradients are correct with a numerical gradient checker. I'll take care of this if @eacharles can point me to a robust way to call the functions directly in python.

Otherwise I'll have to write it in C++ and no one wants that.

**Comment by jasercion** [comment]
Superseded by https://github.com/fermi-lat/ScienceTools/issues/26.  Post all further updates there.
